src/librustc/middle/resolve_lifetime.rs

```diff
@@ -340,15 +340,18 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
                 } else {
                     0
                 };
-                let lifetimes = generics.lifetimes.iter().map(|def| {
-                    Region::early(self.hir_map, &mut index, def)
-                }).collect();
+                let lifetimes = generics.lifetimes()
+                    .map(|def| Region::early(self.hir_map, &mut index, def))
+                    .collect();
                 let scope = Scope::Binder {
                     lifetimes,
                     s: ROOT_SCOPE
                 };
                 self.with(scope, |old_scope, this| {
-                    this.check_lifetime_defs(old_scope, &generics.lifetimes);
+                    this.check_lifetime_defs(
+                        old_scope,
+                        &generics.lifetimes().cloned().collect::<Vec<_>>(),
```

Comment:
Should avoid doing this cloning, preferably just pass in the `&[GenericParam]`.

Reply:
`generics.lifetimes().collect::<Vec<_>>()[..]` is a `&[&GenericParam]` though, is there a way of converting that or would I have to make `check_lifetime_defs` generic over `Iterator<Item = &GenericParam>`?

Reply:
I'd rename `check_lifetime_defs` to `check_lifetime_params` and have it take the entire `&[GenericParam]`, and use `.lifetimes()` inside itself.